[GPU Process] Introduce RenderingResource and make it the base of DecomposedGlyphs and NativeImage

shallawa opened the pull request:
#### 11e07c42d4bfe42c167bdf3ec290d1db6e54694d
<pre>
[GPU Process] Introduce RenderingResource and make it the base of DecomposedGlyphs and NativeImage
 <a href="https://bugs.webkit.org/show_bug.cgi?id=220943">https://bugs.webkit.org/show_bug.cgi?id=220943</a>
 &lt;rdar://problem/73843739&gt;

Reviewed by Darin Adler.

Move smart pointer support and RenderingResourceIdentifier and observers&apos;
management to the new class RenderingResource.

Make DecomposedGlyphs and NativeImage inherit RenderingResource so they can
share this code instead of duplicating it.

Rename releaseRemoteResource() to releaseRenderingResource().

Make DecomposedGlyphs be destructible on any thread and not necessarily on
the main thread. I do not think there was a reason for making it
WTF::DestructionThread::Main from the beginning.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/DecomposedGlyphs.cpp:
(WebCore::DecomposedGlyphs::DecomposedGlyphs):
(WebCore::m_renderingResourceIdentifier): Deleted.
(WebCore::DecomposedGlyphs::~DecomposedGlyphs): Deleted.
* Source/WebCore/platform/graphics/DecomposedGlyphs.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::~NativeImage): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/RenderingResource.h: Added.
(WebCore::RenderingResource::~RenderingResource):
(WebCore::RenderingResource::renderingResourceIdentifier const):
(WebCore::RenderingResource::addObserver):
(WebCore::RenderingResource::removeObserver):
(WebCore::RenderingResource::RenderingResource):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseRenderingResource):
(WebKit::RemoteRenderingBackend::releaseRenderingResourceWithQualifiedIdentifier):
(WebKit::RemoteRenderingBackend::releaseResource): Deleted.
(WebKit::RemoteRenderingBackend::releaseResourceWithQualifiedIdentifier): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::releaseRenderingResource):
(WebKit::RemoteResourceCache::releaseResource): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::releaseRenderingResource):
(WebKit::RemoteRenderingBackendProxy::releaseRemoteResource): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::releaseImageBuffer):
(WebKit::RemoteResourceCacheProxy::releaseRenderingResource):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
(WebKit::RemoteResourceCacheProxy::releaseNativeImage): Deleted.
(WebKit::RemoteResourceCacheProxy::releaseDecomposedGlyphs): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: <a href="https://commits.webkit.org/262021@main">https://commits.webkit.org/262021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28194f8a0f2ea915e2540e437ec393cca153bdef

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/192 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/204 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/257 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/173 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/193 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/351 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/150 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/189 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/211 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/187 "Build was cancelled. Recent messages:OS: Monterey (12.6), Xcode: 14.0; Skipping applying patch since patch_id isn't provided; Checked out pull request; 39 api tests failed or timed out; 39 api tests failed or timed out; Compiled WebKit (warnings); run-api-tests-without-change (cancelled)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/184 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/283 "Build was cancelled. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; layout-tests (exception)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/166 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/193 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/452 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/182 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/189 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/71 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/187 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -141,15 +141,11 @@ void RemoteResourceCacheProxy::recordDecomposedGlyphsUse(DecomposedGlyphs& decom
-    ASSERT_UNUSED(success, success);
-
-    m_remoteRenderingBackendProxy.releaseRemoteResource(renderingResourceIdentifier);
+    bool removed = m_nativeImages.remove(renderingResourceIdentifier) || m_decomposedGlyphs.remove(renderingResourceIdentifier);
```
I kind of don't like looking through all of the `HashMap`s to find the one that contains the resource.

Trying to think of some clever way to continue to have separate functions for releasing. Would this work? Have

* `RenderingResource` be generic, and the type parameter can be `NativeImage`, `DecomposedGlyphs`, etc.
* `NativeImage` inherit from `RenderingResource<NativeImage>`, etc.
* `RemoteResourceCacheProxy` take a reference to the object to release in its `releaseRenderingResource` functions, instead of its ID
* `RenderingResource<T>::Observer::release` call `RemoteResourceCacheProxy::release` with the object to release

heycam replied:
Though I guess this is already how it works in `RemoteResourceCache`. If you want to file an issue for a followup to avoid this in both places, that's OK.

shallawa replied:
I agree. Another solution is to refactor `QualifiedResourceHeap` such that it can be used by `RemoteResourceCacheProxy` as it is used right now by `RemoteResourceCache`.

shallawa committed:
[GPU Process] Introduce RenderingResource and make it the base of DecomposedGlyphs and NativeImage
 https://bugs.webkit.org/show_bug.cgi?id=220943
 <rdar://problem/73843739>

Reviewed by Darin Adler.

Move smart pointer support and RenderingResourceIdentifier and observers'
management to the new class RenderingResource.

Make DecomposedGlyphs and NativeImage inherit RenderingResource so they can
share this code instead of duplicating it.

Rename releaseRemoteResource() to releaseRenderingResource().

Make DecomposedGlyphs be destructible on any thread and not necessarily on
the main thread. I do not think there was a reason for making it
WTF::DestructionThread::Main from the beginning.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/DecomposedGlyphs.cpp:
(WebCore::DecomposedGlyphs::DecomposedGlyphs):
(WebCore::m_renderingResourceIdentifier): Deleted.
(WebCore::DecomposedGlyphs::~DecomposedGlyphs): Deleted.
* Source/WebCore/platform/graphics/DecomposedGlyphs.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::NativeImage):
(WebCore::NativeImage::~NativeImage): Deleted.
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/RenderingResource.h: Added.
(WebCore::RenderingResource::~RenderingResource):
(WebCore::RenderingResource::renderingResourceIdentifier const):
(WebCore::RenderingResource::addObserver):
(WebCore::RenderingResource::removeObserver):
(WebCore::RenderingResource::RenderingResource):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseRenderingResource):
(WebKit::RemoteRenderingBackend::releaseRenderingResourceWithQualifiedIdentifier):
(WebKit::RemoteRenderingBackend::releaseResource): Deleted.
(WebKit::RemoteRenderingBackend::releaseResourceWithQualifiedIdentifier): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::releaseRenderingResource):
(WebKit::RemoteResourceCache::releaseResource): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::releaseRenderingResource):
(WebKit::RemoteRenderingBackendProxy::releaseRemoteResource): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::releaseImageBuffer):
(WebKit::RemoteResourceCacheProxy::releaseRenderingResource):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
(WebKit::RemoteResourceCacheProxy::releaseNativeImage): Deleted.
(WebKit::RemoteResourceCacheProxy::releaseDecomposedGlyphs): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: https://commits.webkit.org/262021@main
